File: internal/record/codec.go
package record

import (
	"encoding/binary"
	"math"
	"sort"
)

// EncodeNodePayload encodes a node as a deterministic MessagePack map for WAL/data payloads.
func EncodeNodePayload(n Node) ([]byte, error) {
	if err := n.ValidateForWrite(); err != nil {
		return nil, err
	}
	m := map[string]any{"type": n.Type, "title": n.Title, "created_at": uint64(n.CreatedAt), "updated_at": uint64(n.UpdatedAt)}
	if n.Body != "" {
		m["body"] = n.Body
	}
	if n.Meta != nil && len(n.Meta) > 0 {
		m["meta"] = n.Meta
	}
	if n.Tags != nil && len(n.Tags) > 0 {
		a := make([]any, len(n.Tags))
		for i, v := range n.Tags {
			a[i] = v
		}
		m["tags"] = a
	}
	if n.Version != 0 && n.Version != 1 {
		m["version"] = uint64(n.Version)
	}
	return encodeMsgpack(m)
}

// DecodeNodePayload decodes a MessagePack node map and applies AKG read defaults.
func DecodeNodePayload(b []byte) (Node, error) {
	v, n, err := decodeMsgpack(b)
	if err != nil || n != len(b) {
		return Node{}, ErrInvalidPayload
	}
	m, ok := v.(map[string]any)
	if !ok {
		return Node{}, ErrInvalidPayload
	}
	typ, ok := m["type"].(string)
	if !ok || typ == "" {
		return Node{}, ErrMissingRequiredField
	}
	title, ok := m["title"].(string)
	if !ok || title == "" {
		return Node{}, ErrMissingRequiredField
	}
	node := Node{Type: typ, Title: title}
	if s, ok := m["body"].(string); ok {
		node.Body = s
	}
	if u, ok := asUint(m["created_at"]); ok {
		node.CreatedAt = TimestampMicros(u)
	}
	if u, ok := asUint(m["updated_at"]); ok {
		node.UpdatedAt = TimestampMicros(u)
	}
	if u, ok := asUint(m["version"]); ok {
		node.Version = Version(u)
	}
	if mm, ok := m["meta"].(map[string]any); ok {
		node.Meta = mm
	}
	if arr, ok := m["tags"].([]any); ok {
		node.Tags = make([]string, 0, len(arr))
		for _, x := range arr {
			s, ok := x.(string)
			if !ok {
				return Node{}, ErrInvalidPayload
			}
			node.Tags = append(node.Tags, s)
		}
	}
	node.ApplyReadDefaults()
	return node, nil
}

func EncodeEdgePayload(e Edge) ([]byte, error) {
	if err := e.ValidateForWrite(); err != nil {
		return nil, err
	}
	m := map[string]any{"from_node": string(e.FromNode), "to_node": string(e.ToNode), "relation": string(e.Relation), "created_at": uint64(e.CreatedAt), "updated_at": uint64(e.UpdatedAt)}
	if e.Strength != 0 {
		m["strength"] = e.Strength
	}
	if e.Confidence != nil {
		m["confidence"] = *e.Confidence
	}
	if e.Meta != nil && len(e.Meta) > 0 {
		m["meta"] = e.Meta
	}
	if e.Version != 0 && e.Version != 1 {
		m["version"] = uint64(e.Version)
	}
	return encodeMsgpack(m)
}

func DecodeEdgePayload(b []byte) (Edge, error) {
	v, n, err := decodeMsgpack(b)
	if err != nil || n != len(b) {
		return Edge{}, ErrInvalidPayload
	}
	m, ok := v.(map[string]any)
	if !ok {
		return Edge{}, ErrInvalidPayload
	}
	from, ok := m["from_node"].(string)
	if !ok || from == "" {
		return Edge{}, ErrMissingRequiredField
	}
	to, ok := m["to_node"].(string)
	if !ok || to == "" {
		return Edge{}, ErrMissingRequiredField
	}
	rel, ok := m["relation"].(string)
	if !ok || rel == "" {
		return Edge{}, ErrMissingRequiredField
	}
	edge := Edge{FromNode: NodeID(from), ToNode: NodeID(to), Relation: Relation(rel)}
	if f, ok := asFloat(m["strength"]); ok {
		edge.Strength = f
	}
	if m["confidence"] != nil {
		if f, ok := asFloat(m["confidence"]); ok {
			edge.Confidence = &f
		}
	}
	if u, ok := asUint(m["created_at"]); ok {
		edge.CreatedAt = TimestampMicros(u)
	}
	if u, ok := asUint(m["updated_at"]); ok {
		edge.UpdatedAt = TimestampMicros(u)
	}
	if u, ok := asUint(m["version"]); ok {
		edge.Version = Version(u)
	}
	if mm, ok := m["meta"].(map[string]any); ok {
		edge.Meta = mm
	}
	edge.ApplyReadDefaults()
	return edge, nil
}

func EncodeNodeDeletePayload(d NodeDelete) ([]byte, error) {
	if err := d.ValidateForWrite(); err != nil {
		return nil, err
	}
	return encodeMsgpack(d.Map())
}
func EncodeEdgeDeletePayload(d EdgeDelete) ([]byte, error) {
	if err := d.ValidateForWrite(); err != nil {
		return nil, err
	}
	return encodeMsgpack(d.Map())
}
func DecodeNodeDeletePayload(b []byte) (NodeDelete, error) {
	v, n, err := decodeMsgpack(b)
	if err != nil || n != len(b) {
		return NodeDelete{}, ErrInvalidPayload
	}
	m, ok := v.(map[string]any)
	if !ok {
		return NodeDelete{}, ErrInvalidPayload
	}
	return NodeDeleteFromMap(m)
}
func DecodeEdgeDeletePayload(b []byte) (EdgeDelete, error) {
	v, n, err := decodeMsgpack(b)
	if err != nil || n != len(b) {
		return EdgeDelete{}, ErrInvalidPayload
	}
	m, ok := v.(map[string]any)
	if !ok {
		return EdgeDelete{}, ErrInvalidPayload
	}
	return EdgeDeleteFromMap(m)
}

func encodeMsgpack(v any) ([]byte, error) {
	var out []byte
	err := appendMsgpack(&out, v)
	return out, err
}

func appendMsgpack(out *[]byte, v any) error {
	switch x := v.(type) {
	case nil:
		*out = append(*out, 0xc0)
	case bool:
		if x {
			*out = append(*out, 0xc3)
		} else {
			*out = append(*out, 0xc2)
		}
	case string:
		appendStr(out, x)
	case uint64:
		*out = append(*out, 0xcf)
		var b [8]byte
		binary.BigEndian.PutUint64(b[:], x)
		*out = append(*out, b[:]...)
	case uint32:
		return appendMsgpack(out, uint64(x))
	case int:
		return appendMsgpack(out, uint64(x))
	case float64:
		*out = append(*out, 0xcb)
		var b [8]byte
		binary.BigEndian.PutUint64(b[:], math.Float64bits(x))
		*out = append(*out, b[:]...)
	case []any:
		appendArrayHeader(out, len(x))
		for _, e := range x {
			if err := appendMsgpack(out, e); err != nil {
				return err
			}
		}
	case map[string]any:
		appendMapHeader(out, len(x))
		keys := make([]string, 0, len(x))
		for k := range x {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		for _, k := range keys {
			appendStr(out, k)
			if err := appendMsgpack(out, x[k]); err != nil {
				return err
			}
		}
	default:
		return ErrInvalidPayload
	}
	return nil
}

func appendStr(out *[]byte, s string) {
	l := len(s)
	if l < 32 {
		*out = append(*out, byte(0xa0|l))
	} else {
		*out = append(*out, 0xdb)
		var b [4]byte
		binary.BigEndian.PutUint32(b[:], uint32(l))
		*out = append(*out, b[:]...)
	}
	*out = append(*out, s...)
}
func appendMapHeader(out *[]byte, l int) {
	if l < 16 {
		*out = append(*out, byte(0x80|l))
	} else {
		*out = append(*out, 0xde, byte(l>>8), byte(l))
	}
}
func appendArrayHeader(out *[]byte, l int) {
	if l < 16 {
		*out = append(*out, byte(0x90|l))
	} else {
		*out = append(*out, 0xdc, byte(l>>8), byte(l))
	}
}

func decodeMsgpack(b []byte) (any, int, error) {
	if len(b) == 0 {
		return nil, 0, ErrInvalidPayload
	}
	c := b[0]
	if c <= 0x7f {
		return uint64(c), 1, nil
	}
	if c >= 0xa0 && c <= 0xbf {
		l := int(c & 0x1f)
		if len(b) < 1+l {
			return nil, 0, ErrInvalidPayload
		}
		return string(b[1 : 1+l]), 1 + l, nil
	}
	if c >= 0x90 && c <= 0x9f {
		return decodeArray(b, 1, int(c&0x0f))
	}
	if c >= 0x80 && c <= 0x8f {
		return decodeMap(b, 1, int(c&0x0f))
	}
	switch c {
	case 0xc0:
		return nil, 1, nil
	case 0xc2:
		return false, 1, nil
	case 0xc3:
		return true, 1, nil
	case 0xcc:
		if len(b) < 2 {
			return nil, 0, ErrInvalidPayload
		}
		return uint64(b[1]), 2, nil
	case 0xcd:
		if len(b) < 3 {
			return nil, 0, ErrInvalidPayload
		}
		return uint64(binary.BigEndian.Uint16(b[1:3])), 3, nil
	case 0xce:
		if len(b) < 5 {
			return nil, 0, ErrInvalidPayload
		}
		return uint64(binary.BigEndian.Uint32(b[1:5])), 5, nil
	case 0xcf:
		if len(b) < 9 {
			return nil, 0, ErrInvalidPayload
		}
		return binary.BigEndian.Uint64(b[1:9]), 9, nil
	case 0xcb:
		if len(b) < 9 {
			return nil, 0, ErrInvalidPayload
		}
		return math.Float64frombits(binary.BigEndian.Uint64(b[1:9])), 9, nil
	case 0xd9:
		if len(b) < 2 {
			return nil, 0, ErrInvalidPayload
		}
		l := int(b[1])
		if len(b) < 2+l {
			return nil, 0, ErrInvalidPayload
		}
		return string(b[2 : 2+l]), 2 + l, nil
	case 0xdb:
		if len(b) < 5 {
			return nil, 0, ErrInvalidPayload
		}
		l := int(binary.BigEndian.Uint32(b[1:5]))
		if len(b) < 5+l {
			return nil, 0, ErrInvalidPayload
		}
		return string(b[5 : 5+l]), 5 + l, nil
	case 0xdc:
		if len(b) < 3 {
			return nil, 0, ErrInvalidPayload
		}
		return decodeArray(b, 3, int(binary.BigEndian.Uint16(b[1:3])))
	case 0xde:
		if len(b) < 3 {
			return nil, 0, ErrInvalidPayload
		}
		return decodeMap(b, 3, int(binary.BigEndian.Uint16(b[1:3])))
	default:
		return nil, 0, ErrInvalidPayload
	}
}

func decodeArray(b []byte, off, l int) (any, int, error) {
	arr := make([]any, l)
	pos := off
	for i := 0; i < l; i++ {
		v, n, err := decodeMsgpack(b[pos:])
		if err != nil {
			return nil, 0, err
		}
		arr[i] = v
		pos += n
	}
	return arr, pos, nil
}
func decodeMap(b []byte, off, l int) (any, int, error) {
	m := make(map[string]any, l)
	pos := off
	for i := 0; i < l; i++ {
		k, n, err := decodeMsgpack(b[pos:])
		if err != nil {
			return nil, 0, err
		}
		pos += n
		ks, ok := k.(string)
		if !ok {
			return nil, 0, ErrInvalidPayload
		}
		v, n, err := decodeMsgpack(b[pos:])
		if err != nil {
			return nil, 0, err
		}
		pos += n
		m[ks] = v
	}
	return m, pos, nil
}
func asUint(v any) (uint64, bool) { u, ok := v.(uint64); return u, ok }
func asFloat(v any) (float64, bool) {
	switch x := v.(type) {
	case float64:
		return x, true
	case uint64:
		return float64(x), true
	default:
		return 0, false
	}
}

File: sdk/akg-ts/src/internal/codec.ts
import { InvalidInputError, MissingRequiredFieldError } from '../errors.js';
import { encodeMsgpack, decodeMsgpackFull, type MsgpackValue } from './msgpack.js';

export interface CoreNode {
  type: string;
  title: string;
  body: string;
  meta: Record<string, unknown>;
  tags: string[];
  createdAt: bigint;
  updatedAt: bigint;
  version: number;
}

export interface CoreEdge {
  fromType: string;
  fromNode: string;
  toType: string;
  toNode: string;
  relation: string;
  strength: number;
  confidence: number | null;
  meta: Record<string, unknown>;
  createdAt: bigint;
  updatedAt: bigint;
  version: number;
}

export interface NodePut {
  id: string;
  node: CoreNode;
}

export interface NodeDelete {
  type: string;
  id: string;
}

export interface EdgeDelete {
  fromType: string;
  fromNode: string;
  relation: string;
  toType: string;
  toNode: string;
}

function asUint(v: MsgpackValue): bigint | null {
  if (typeof v === 'number' && Number.isInteger(v) && v >= 0) return BigInt(v);
  if (typeof v === 'bigint' && v >= 0n) return v;
  return null;
}

function toMsgpackMeta(meta: Record<string, unknown>): MsgpackValue {
  const m: Record<string, MsgpackValue> = {};
  for (const [k, v] of Object.entries(meta)) {
    m[k] = toMsgpackValue(v);
  }
  return m;
}

function toMsgpackValue(v: unknown): MsgpackValue {
  if (v === null || v === undefined) return null;
  if (typeof v === 'boolean') return v;
  if (typeof v === 'string') return v;
  if (typeof v === 'number') return v;
  if (typeof v === 'bigint') return v;
  if (Array.isArray(v)) return v.map(toMsgpackValue);
  if (typeof v === 'object') {
    const m: Record<string, MsgpackValue> = {};
    for (const [k, val] of Object.entries(v as Record<string, unknown>)) {
      m[k] = toMsgpackValue(val);
    }
    return m;
  }
  throw new InvalidInputError('unsupported meta value type');
}

function fromMsgpackMeta(v: MsgpackValue): Record<string, unknown> {
  if (v === null || v === undefined) return {};
  if (typeof v !== 'object' || Array.isArray(v)) throw new InvalidInputError('invalid meta');
  return v as Record<string, unknown>;
}

export function encodeNodePayload(n: CoreNode): Uint8Array {
  if (!n.type || !n.title) throw new MissingRequiredFieldError();
  const m: Record<string, MsgpackValue> = {
    type: n.type,
    title: n.title,
    created_at: n.createdAt,
    updated_at: n.updatedAt,
  };
  if (n.body) m.body = n.body;
  if (Object.keys(n.meta).length > 0) m.meta = toMsgpackMeta(n.meta);
  if (n.tags.length > 0) m.tags = n.tags;
  if (n.version !== 0 && n.version !== 1) m.version = BigInt(n.version);
  return encodeMsgpack(m);
}

export function encodeNodePutPayload(p: NodePut): Uint8Array {
  if (!p.id) throw new MissingRequiredFieldError();
  if (!p.node.type || !p.node.title) throw new MissingRequiredFieldError();
  const m: Record<string, MsgpackValue> = {
    id: p.id,
    type: p.node.type,
    title: p.node.title,
    created_at: p.node.createdAt,
    updated_at: p.node.updatedAt,
  };
  if (p.node.body) m.body = p.node.body;
  if (Object.keys(p.node.meta).length > 0) m.meta = toMsgpackMeta(p.node.meta);
  if (p.node.tags.length > 0) m.tags = p.node.tags;
  if (p.node.version !== 0 && p.node.version !== 1) m.version = BigInt(p.node.version);
  return encodeMsgpack(m);
}

export function decodeNodePayload(b: Uint8Array): CoreNode {
  let v: MsgpackValue;
  try {
    v = decodeMsgpackFull(b);
  } catch {
    throw new InvalidInputError('invalid data payload');
  }
  if (typeof v !== 'object' || Array.isArray(v) || v === null) throw new InvalidInputError('invalid data payload');
  const m = v as Record<string, MsgpackValue>;
  const type = m.type;
  if (typeof type !== 'string' || !type) throw new MissingRequiredFieldError('node type missing');
  const title = m.title;
  if (typeof title !== 'string' || !title) throw new MissingRequiredFieldError('node title missing');
  const node: CoreNode = {
    type,
    title,
    body: typeof m.body === 'string' ? m.body : '',
    meta: m.meta ? fromMsgpackMeta(m.meta) : {},
    tags: [],
    createdAt: 0n,
    updatedAt: 0n,
    version: 1,
  };
  const ts = asUint(m.created_at);
  if (ts !== null) node.createdAt = ts;
  const tu = asUint(m.updated_at);
  if (tu !== null) node.updatedAt = tu;
  const ver = asUint(m.version);
  if (ver !== null) node.version = Number(ver);
  if (node.version === 0) node.version = 1;
  if (Array.isArray(m.tags)) {
    for (const t of m.tags) {
      if (typeof t !== 'string') throw new InvalidInputError('invalid data payload: tag not a string');
      node.tags.push(t);
    }
  }
  return node;
}

export function decodeNodePutPayload(b: Uint8Array): NodePut {
  let v: MsgpackValue;
  try {
    v = decodeMsgpackFull(b);
  } catch {
    throw new InvalidInputError('invalid wal payload');
  }
  if (typeof v !== 'object' || Array.isArray(v) || v === null) throw new InvalidInputError('invalid wal payload');
  const m = v as Record<string, MsgpackValue>;
  const id = m.id;
  if (typeof id !== 'string' || !id) throw new MissingRequiredFieldError('node id missing');
  const node = decodeNodeFromMap(m);
  return { id, node };
}

function decodeNodeFromMap(m: Record<string, MsgpackValue>): CoreNode {
  const type = m.type;
  if (typeof type !== 'string' || !type) throw new MissingRequiredFieldError('node type missing');
  const title = m.title;
  if (typeof title !== 'string' || !title) throw new MissingRequiredFieldError('node title missing');
  const node: CoreNode = {
    type,
    title,
    body: typeof m.body === 'string' ? m.body : '',
    meta: m.meta ? fromMsgpackMeta(m.meta) : {},
    tags: [],
    createdAt: 0n,
    updatedAt: 0n,
    version: 1,
  };
  const ts = asUint(m.created_at);
  if (ts !== null) node.createdAt = ts;
  const tu = asUint(m.updated_at);
  if (tu !== null) node.updatedAt = tu;
  const ver = asUint(m.version);
  if (ver !== null) node.version = Number(ver);
  if (node.version === 0) node.version = 1;
  if (Array.isArray(m.tags)) {
    for (const t of m.tags) {
      if (typeof t !== 'string') throw new InvalidInputError('invalid payload: tag not a string');
      node.tags.push(t);
    }
  }
  return node;
}

export function encodeNodeDeletePayload(d: NodeDelete): Uint8Array {
  if (!d.type || !d.id) throw new MissingRequiredFieldError();
  return encodeMsgpack({ type: d.type, id: d.id });
}

export function decodeNodeDeletePayload(b: Uint8Array): NodeDelete {
  let v: MsgpackValue;
  try {
    v = decodeMsgpackFull(b);
  } catch {
    throw new InvalidInputError('invalid wal payload');
  }
  if (typeof v !== 'object' || Array.isArray(v) || v === null) throw new InvalidInputError('invalid wal payload');
  const m = v as Record<string, MsgpackValue>;
  const type = m.type;
  if (typeof type !== 'string' || !type) throw new MissingRequiredFieldError('node type missing');
  const id = m.id;
  if (typeof id !== 'string' || !id) throw new MissingRequiredFieldError('node id missing');
  return { type, id };
}

export function encodeEdgePayload(e: CoreEdge): Uint8Array {
  if (!e.fromType || !e.fromNode || !e.toType || !e.toNode || !e.relation) throw new MissingRequiredFieldError();
  const m: Record<string, MsgpackValue> = {
    from_node_type: e.fromType,
    from_node: e.fromNode,
    to_node_type: e.toType,
    to_node: e.toNode,
    relation: e.relation,
    strength: e.strength,
    created_at: e.createdAt,
    updated_at: e.updatedAt,
  };
  if (e.confidence !== null) m.confidence = e.confidence;
  if (Object.keys(e.meta).length > 0) m.meta = toMsgpackMeta(e.meta);
  if (e.version !== 0 && e.version !== 1) m.version = BigInt(e.version);
  return encodeMsgpack(m);
}

export function decodeEdgePayload(b: Uint8Array): CoreEdge {
  let v: MsgpackValue;
  try {
    v = decodeMsgpackFull(b);
  } catch {
    throw new InvalidInputError('invalid data payload');
  }
  if (typeof v !== 'object' || Array.isArray(v) || v === null) throw new InvalidInputError('invalid data payload');
  const m = v as Record<string, MsgpackValue>;
  return decodeEdgeFromMap(m);
}

export function decodeEdgePutPayload(b: Uint8Array): CoreEdge {
  let v: MsgpackValue;
  try {
    v = decodeMsgpackFull(b);
  } catch {
    throw new InvalidInputError('invalid wal payload');
  }
  if (typeof v !== 'object' || Array.isArray(v) || v === null) throw new InvalidInputError('invalid wal payload');
  const m = v as Record<string, MsgpackValue>;
  return decodeEdgeFromMap(m);
}

function decodeEdgeFromMap(m: Record<string, MsgpackValue>): CoreEdge {
  const fromType = m.from_node_type;
  if (typeof fromType !== 'string' || !fromType) throw new MissingRequiredFieldError('from_node_type missing');
  const fromNode = m.from_node;
  if (typeof fromNode !== 'string' || !fromNode) throw new MissingRequiredFieldError('from_node missing');
  const toType = m.to_node_type;
  if (typeof toType !== 'string' || !toType) throw new MissingRequiredFieldError('to_node_type missing');
  const toNode = m.to_node;
  if (typeof toNode !== 'string' || !toNode) throw new MissingRequiredFieldError('to_node missing');
  const relation = m.relation;
  if (typeof relation !== 'string' || !relation) throw new MissingRequiredFieldError('relation missing');

  const edge: CoreEdge = {
    fromType,
    fromNode,
    toType,
    toNode,
    relation,
    strength: 0.5,
    confidence: null,
    meta: {},
    createdAt: 0n,
    updatedAt: 0n,
    version: 1,
  };
  if (typeof m.strength === 'number') edge.strength = m.strength;
  if ('confidence' in m) {
    if (m.confidence === null) {
      edge.confidence = null;
    } else if (typeof m.confidence === 'number') {
      edge.confidence = m.confidence;
    } else {
      throw new InvalidInputError('invalid payload: confidence not a number');
    }
  }
  const ts = asUint(m.created_at);
  if (ts !== null) edge.createdAt = ts;
  const tu = asUint(m.updated_at);
  if (tu !== null) edge.updatedAt = tu;
  const ver = asUint(m.version);
  if (ver !== null) edge.version = Number(ver);
  if (edge.version === 0) edge.version = 1;
  if (m.meta) edge.meta = fromMsgpackMeta(m.meta);
  return edge;
}

export function encodeEdgeDeletePayload(d: EdgeDelete): Uint8Array {
  if (!d.fromType || !d.fromNode || !d.relation || !d.toType || !d.toNode) throw new MissingRequiredFieldError();
  return encodeMsgpack({
    from_node_type: d.fromType,
    from_node: d.fromNode,
    relation: d.relation,
    to_node_type: d.toType,
    to_node: d.toNode,
  });
}

export function decodeEdgeDeletePayload(b: Uint8Array): EdgeDelete {
  let v: MsgpackValue;
  try {
    v = decodeMsgpackFull(b);
  } catch {
    throw new InvalidInputError('invalid wal payload');
  }
  if (typeof v !== 'object' || Array.isArray(v) || v === null) throw new InvalidInputError('invalid wal payload');
  const m = v as Record<string, MsgpackValue>;
  const fromType = m.from_node_type;
  if (typeof fromType !== 'string' || !fromType) throw new MissingRequiredFieldError('from_node_type missing');
  const fromNode = m.from_node;
  if (typeof fromNode !== 'string' || !fromNode) throw new MissingRequiredFieldError('from_node missing');
  const relation = m.relation;
  if (typeof relation !== 'string' || !relation) throw new MissingRequiredFieldError('relation missing');
  const toType = m.to_node_type;
  if (typeof toType !== 'string' || !toType) throw new MissingRequiredFieldError('to_node_type missing');
  const toNode = m.to_node;
  if (typeof toNode !== 'string' || !toNode) throw new MissingRequiredFieldError('to_node missing');
  return { fromType, fromNode, relation, toType, toNode };
}

export function validateWALPayload(op: number, payload: Uint8Array): void {
  switch (op) {
    case 0x01: decodeNodePutPayload(payload); break;
    case 0x02: decodeNodeDeletePayload(payload); break;
    case 0x03: decodeEdgePutPayload(payload); break;
    case 0x04: decodeEdgeDeletePayload(payload); break;
    case 0x05: if (payload.length !== 0) throw new InvalidInputError('invalid wal record'); break;
    default: throw new InvalidInputError('unknown wal operation');
  }
}
